File: C/password_gen.c
//
// Created by spide on 4/17/2026.
//

#include "password_gen.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

//private vars
int password_length = 0;
char password_type;
char* password;

//getting the options the user selected
void setting_length(int length) { //setting the length of the password
    password_length = length;
}

void password_options(char option) { //setting the option that the user got
    password_type = option;
}

//opening files
char** file_opener(const char* path, int* out_count) {
    FILE* f = fopen(path, "r");
    //if the file can not be opened
    if(!f) {
        printf("error opening the file: %s. Please check the spelling of the file name or if the configuration is correct.", path);
        *out_count = 0;
        return NULL;
    }

    char** lines = NULL;
    int count = 0;
    char buffer[256];

    //opens the file and place it on a section
    while (fgets(buffer,sizeof(buffer),f)) {
        buffer[strlen(buffer)-1] = '\0';

        lines = realloc(lines, sizeof(char*) * (count + 1));
        lines[count] = malloc(strlen(buffer) + 1);
        strcpy(lines[count], buffer);

        count++;
    }

    //closing the file
    fclose(f);
    *out_count = count;
    return lines;
}

//getting a random item
char random_item(char** list, int count) {
    int index = rand() % count;
    return list[index][0];
}

//adding items at the end of the password
void fill_password(char* password, int length, char** list, int count) {
    for (int i = 0; i < length; i++) {
        password[i] = random_item(list, count);
    }
    password[length] = '\0';
}


//main password system
void making_password() {
    //seed
    srand(time(NULL));

    //private vars
    int count = 0;
    char** list = NULL;
    password = malloc(password_length + 1);


    //checking what the user got
    switch (password_type) {
        case 'a':
            printf("words only");

            //gets the alphabet text file
            list = file_opener("data/alphabet.txt", &count);

            //places a random letter into the password
            fill_password(password, password_length, list, count);

            //prints out the password
            printf("Password: %s\n", password);
            break;
        case 'b':
            printf("numbers only");

            //loads the list
            list = file_opener("data/numbers.txt", &count);

            //prints out the password
            fill_password(password, password_length, list, count);

            //prints the password
            printf("Password: %s\n", password);
            break;
        case 'c':
            printf("spcial charcaters only");

            //gets the file
            list = file_opener("data/special.txt", &count);

            //prints the password
            fill_password(password, password_length, list, count);
            printf("Password: %s\n", password);

            break;
        case 'd':
            printf("all options");
            break;
    }
}

File: Go/password_gen.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"strings"
	"time"
)

// vars
var password_length_private int
var password_options_private string
var password string = ""

// setters
func settingLength(length int) {
	password_length_private = length
}
func settingOptions(options string) {
	password_options_private = options
}

// opening a file
func openingFile(filename string) ([]string, error) {
	data, err := os.ReadFile(filename)
	if err != nil {
		return nil, err
	}

	// convert file contents into a slice of strings
	lines := strings.Split(string(data), "\n")

	// trim whitespace + remove empty lines
	var result []string
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line != "" {
			result = append(result, line)
		}
	}

	return result, nil
}

// making the new password
func newPassword(pass string, length int, list []string) string {
	var index int

	for i := 0; i < length; i++ {
		index = rand.Intn(len(list))
		pass = pass + list[index]
	}
	return pass
}

// making a simple copier
func copyNewPassword(pass string, length int, fileName string) string {
	//opens a file
	items, err := openingFile(fileName)

	//if there is an error while opening it, it will yell
	if err != nil {
		fmt.Println(err)
	}

	pass = newPassword(pass, length, items)

	return pass
}

// main function
func mainGen() {
	//setting the seed
	rand.Seed(time.Now().UnixNano())

	switch password_options_private {
	case "a":
		fmt.Println("letters only")

		password = copyNewPassword(password, password_length_private, "../data/alphabet.txt")

		fmt.Println("Your password: ", password)

		break
	case "b":
		fmt.Println("numbers only")

		password = copyNewPassword(password, password_length_private, "../data/numbers.txt")
		fmt.Println("Your password: ", password)

		break
	case "c":
		fmt.Println("symbols only")

		password = copyNewPassword(password, password_length_private, "../data/special.txt")
		fmt.Println("Your password: ", password)

		break
	case "d":
		fmt.Println("all options")

		//opening the files

		//opening the alphabet file
		leters, letterErr := openingFile("../data/alphabet.txt")
		if letterErr != nil {
			fmt.Println(letterErr)
		}

		//opening the number files
		numbers, numberErr := openingFile("../data/numbers.txt")
		if numberErr != nil {
			fmt.Println(numberErr)
		}

		//opening the special chars
		special, specialErr := openingFile("../data/special.txt")
		if specialErr != nil {
			fmt.Println(specialErr)
		}

		for i := 0; i < password_length_private; i++ {
			var randNum int = rand.Intn(3)
			switch randNum {
			case 0:
				password = newPassword(password, 1, leters)
			case 1:
				password = newPassword(password, 1, numbers)
			case 2:
				password = newPassword(password, 1, special)
			}
		}
		fmt.Println(password)

		break
	}
}
